tracer.go

```diff
@@ -0,0 +1,297 @@
+type PBTracer struct {
+	basicTracer
+	w io.WriteCloser
+}
+
+func NewPBTracer(file string) (*PBTracer, error) {
+	return OpenPBTracer(file, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
+}
+
+// OpenPBTracer creates a new PBTracer, with explicit control of OpenFile flags and permissions.
+func OpenPBTracer(file string, flags int, perm os.FileMode) (*PBTracer, error) {
+	f, err := os.OpenFile(file, flags, perm)
+	if err != nil {
+		return nil, err
+	}
+
+	tr := &PBTracer{w: f, basicTracer: basicTracer{ch: make(chan struct{}, 1)}}
+	go tr.doWrite()
+
+	return tr, nil
+}
+
+func (t *PBTracer) doWrite() {
+	var buf []*pb.TraceEvent
+	w := ggio.NewDelimitedWriter(t.w)
+	for {
+		_, ok := <-t.ch
+
+		t.mx.Lock()
+		tmp := t.buf
+		t.buf = buf[:0]
+		buf = tmp
+		t.mx.Unlock()
+
+		for i, evt := range buf {
+			err := w.WriteMsg(evt)
+			if err != nil {
+				log.Errorf("error writing event trace: %s", err.Error())
+			}
+			buf[i] = nil
+		}
+
+		if !ok {
+			t.w.Close()
+			return
+		}
+	}
+}
+
+var _ EventTracer = (*PBTracer)(nil)
+
+const RemoteTracerProtoID = protocol.ID("/libp2p/pubsub/tracer/1.0.0")
+
+// RemoteTracer is a tracer that sends trace events to a remote peer
+type RemoteTracer struct {
+	basicTracer
+	ctx  context.Context
+	host host.Host
+	pi   peer.AddrInfo
+}
+
+// NewRemoteTracer constructs a RemoteTracer, tracing to the peer identified by pi
+func NewRemoteTracer(ctx context.Context, host host.Host, pi peer.AddrInfo) (*RemoteTracer, error) {
+	tr := &RemoteTracer{ctx: ctx, host: host, pi: pi, basicTracer: basicTracer{ch: make(chan struct{}, 1), lossy: true}}
+	go tr.doWrite()
+	return tr, nil
+}
+
+func (t *RemoteTracer) doWrite() {
+	var buf []*pb.TraceEvent
+
+	s, err := t.openStream()
+	if err != nil {
+		log.Errorf("error opening remote tracer stream: %s", err.Error())
+		return
+	}
+
+	var batch pb.TraceEventBatch
+
+	gzipW := gzip.NewWriter(s)
+	w := ggio.NewDelimitedWriter(gzipW)
+
+	for {
+		_, ok := <-t.ch
+
+		// nil out the buffer to gc events when swapping buffers
+		for i := range buf {
+			buf[i] = nil
+		}
+
+		// wait a bit to accumulate a batch
+		time.Sleep(time.Second)
+
+		t.mx.Lock()
+		tmp := t.buf
+		t.buf = buf[:0]
+		buf = tmp
+		t.mx.Unlock()
+
+		if len(buf) == 0 {
+			goto end
+		}
+
+		batch.Batch = buf
+
+		err = w.WriteMsg(&batch)
+		if err != nil {
+			log.Errorf("error writing trace event batch: %s", err)
+			goto end
+		}
+
+		err = gzipW.Flush()
+		if err != nil {
+			log.Errorf("error flushin gzip stream: %s", err)
+			goto end
+		}
+
+	end:
+		if !ok {
+			if err != nil {
+				s.Reset()
+			} else {
+				gzipW.Close()
+				helpers.FullClose(s)
+			}
+			return
+		}
+
+		if err != nil {
+			s.Reset()
+			s, err = t.openStream()
+			if err != nil {
+				log.Errorf("error opening remote tracer stream: %s", err.Error())
+				return
+			}
+
+			gzipW.Reset(s)
+		}
+	}
+}
+
+func (t *RemoteTracer) connect() error {
+	for {
+		ctx, cancel := context.WithTimeout(t.ctx, time.Minute)
```

Comment:
Can't we just call open a stream? We shouldn't have to dial separately.

Reply:
We can, but the host will have to know how to reach the trace peer. The traced is not announcing itself to the dht, so it wouldn't be findable in general.

Reply:
Ah. Shouldn't we add the the addresses to the peerstore on start with an expiration of "never" (https://github.com/libp2p/go-libp2p-core/blob/master/peerstore/peerstore.go#L45)?

Reply:
that would work too; can do.

Reply:
done.